ir/expression.def

```diff
@@ -325,26 +325,52 @@ class TypeNameExpression : Expression {
                          "%1% unexpected type in TypeNameExpression", typeName); }
 }
 
-class Slice : Operation_Ternary {
+abstract Slice : Operation_Ternary {
+    virtual unsigned getH() const = 0;
+    virtual unsigned getL() const = 0;
+    // make a slice, folding slices on slices and slices on constants automatically
+    static Expression make(Expression a, unsigned hi, unsigned lo);
+}
+
+class AbsSlice : Slice {
     precedence = DBPrint::Prec_Postfix;
     stringOp = "[:]";
     toString{ return absl::StrCat(e0->toString().string_view(),
                                   "[", e1->toString().string_view(), ":",
                                        e2->toString().string_view(), "]"); }
     // After type checking e1 and e2 will be constants
-    unsigned getH() const { return e1->to<IR::Constant>()->asUnsigned(); }
-    unsigned getL() const { return e2->to<IR::Constant>()->asUnsigned(); }
-    Slice(Expression a, int hi, int lo)
-    : Operation_Ternary(IR::Type::Bits::get(hi-lo+1), a, new Constant(hi), new Constant(lo)) {}
-    Slice(Util::SourceInfo si, Expression a, int hi, int lo)
-    : Operation_Ternary(si, IR::Type::Bits::get(hi-lo+1), a, new Constant(hi), new Constant(lo)) {}
-    Slice {
+    unsigned getH() const override { return e1->to<IR::Constant>()->asUnsigned(); }
+    unsigned getL() const override { return e2->to<IR::Constant>()->asUnsigned(); }
```

Comment:
Same here, checked cast.

Reply:
These weren't checked casts before, so I was just keeping the same behavior.

Reply:
Yes, but this seems to lead a nullpointer access if the cast does not work.

Reply:
Which has the same net result -- an internal compiler error.

Reply:
True, but checkedTo adds some more output to the error.